### PR TITLE
Add tests for PBFT

### PR DESCRIPTION
**Pull request by bailantaotao**

### CODE REVIEW

File: consensus/pbft/core/testbackend_test.go [hunk 1]
```diff
@@ -0,0 +1,258 @@
...
+	return self.peers
+}
+
+func (self *testSystemBackend) EventMux() *event.TypeMux {
+	return self.events
+}
+
+func (self *testSystemBackend) Send(message []byte) {
+	testLogger.Info("enqueuing a message...", "id", self.ID())
+	self.sys.queuedMessage <- &testMessage{
```
Comment (tailingchen) [inline consensus/pbft/core/testbackend_test.go]: if you'd like to handle target peer outside, and you need to use `pbft.MessageEvent`  here.
```
pbft.MessageEvent{
    ID:      self.ID(),
    Payload: message,
 }
```

File: consensus/pbft/core/testbackend_test.go [hunk 2]
```diff
@@ -0,0 +1,258 @@
...
+	testLogger.Info(fmt.Sprintf("add peer: %d", self.Peers().GetByPublicKey(peerPublicKey).ID()), "id", self.ID())
+}
+
+// Remove a peer
+func (self *testSystemBackend) RemovePeer(peerPublicKey string) {
+	testLogger.Warn("nothing to happen")
+}
+
+// Handle a message from peer
+func (self *testSystemBackend) HandleMsg(peerPublicKey string, data []byte) {
```
Comment (tailingchen) [inline consensus/pbft/core/testbackend_test.go]: we should decode `data ` with `pbft.MessageEvent` type here.

Comment (bailantaotao) [inline consensus/pbft/core/testbackend_test.go]: fixed in https://github.com/getamis/go-ethereum/pull/20/commits/6ef8dd2dc9f987da781a9ec8f25a8848cf0ac938

File: consensus/pbft/core/testbackend_test.go (hunk 1 above)
Comment (bailantaotao) [inline consensus/pbft/core/testbackend_test.go]: fixed in https://github.com/getamis/go-ethereum/pull/20/commits/6ef8dd2dc9f987da781a9ec8f25a8848cf0ac938

File: consensus/pbft/core/testbackend_test.go [hunk 3]
```diff
@@ -182,9 +177,16 @@ func (t *testSystem) run() {
...
 			case queuedMessage := <-t.queuedMessage:
-				testLogger.Info("consuming a queue message...", "msg from", queuedMessage.From)
+				testLogger.Info("consuming a queue message...", "msg from", queuedMessage.ID)
 				for _, backend := range t.backends {
-					go backend.HandleMsg(backend.peers.GetByIndex(queuedMessage.From).PublicKey(), queuedMessage.Message)
+					if backend.ID() != queuedMessage.ID {
+						go backend.HandleMsg(backend.peers.GetByIndex(queuedMessage.ID).PublicKey(), queuedMessage.Payload)
+						continue
+					}
+					go backend.EventMux().Post(pbft.MessageEvent{
```
Comment (tailingchen) [inline consensus/pbft/core/testbackend_test.go]:
```
go backend.EventMux().Post(*queuedMessage)
```

Comment (bailantaotao) [inline consensus/pbft/core/testbackend_test.go]: fixed in 18357a31a3041f060fe1df45cf2b0f22122c017b

**Pull request merged**